### PR TITLE
Add in-memory storage

### PR DESCRIPTION
**Pull request by pvdrz**
# Description of change

Adds the `bee-storage-volatile` crate as an in-memory storage backend.

## Type of change

- Enhancement (a non-breaking change which adds functionality)

## How the change has been tested

Running the `bee-storage-test` suite.

## Change checklist

- [x] I have followed the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that new and existing unit tests pass locally with my changes
- [x] I have updated the CHANGELOG.md, if my changes are significant enough

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -15,7 +15,8 @@ members = [
 	"bee-storage/bee-storage-null",
 	"bee-storage/bee-storage-rocksdb",
 	"bee-storage/bee-storage-sled",
-	"bee-storage/bee-storage-test",
+    "bee-storage/bee-storage-test",
```
Comment (thibault-martinez) [inline Cargo.toml]: align!

File: bee-storage/bee-storage-volatile/src/storage.rs [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+
+impl<T> From<PoisonError<T>> for Error {
+    fn from(_: PoisonError<T>) -> Self {
+        Self::PoisonedLock
+    }
+}
+
+pub(crate) const STORAGE_VERSION: StorageVersion = StorageVersion(0);
+
+/// The sled database.
```
Comment (thibault-martinez) [inline bee-storage/bee-storage-volatile/src/storage.rs]: Not sled

File: bee-storage/bee-storage-volatile/src/storage.rs [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+    milestone::Milestone,
+    prelude::{MilestoneIndex, OutputId, PaddedIndex},
+    Message, MessageId,
+};
+use bee_storage::{
+    access::{Fetch, Insert},
+    backend::StorageBackend,
+    system::{StorageHealth, StorageVersion, System, SYSTEM_HEALTH_KEY, SYSTEM_VERSION_KEY},
+};
+
```
Comment (thibault-martinez) [inline bee-storage/bee-storage-volatile/src/storage.rs]:
```suggestion
```

File: bee-storage/bee-storage-volatile/src/storage.rs [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+use bee_tangle::{
+    metadata::MessageMetadata, solid_entry_point::SolidEntryPoint, unreferenced_message::UnreferencedMessage,
+};
+use thiserror::Error;
```
Comment (thibault-martinez) [inline bee-storage/bee-storage-volatile/src/storage.rs]:
```suggestion

use thiserror::Error;
```

File: bee-storage/bee-storage-volatile/src/storage.rs [hunk 5]
```diff
@@ -0,0 +1,133 @@
+// Copyright 2021 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+//! The sled storage backend.
```
Comment (thibault-martinez) [inline bee-storage/bee-storage-volatile/src/storage.rs]: Not sled

File: bee-storage/bee-storage-volatile/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+// SPDX-License-Identifier: Apache-2.0
+
+//! In-memory Bee storage backend.
+
+#![deny(missing_docs)]
+#![deny(warnings)]
+
+pub mod access;
+pub mod storage;
+mod table;
```
Comment (thibault-martinez) [inline bee-storage/bee-storage-volatile/src/lib.rs]: Can you put mod before then a newline please

File: bee-storage/bee-storage-volatile/Cargo.toml [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+homepage = "https://www.iota.org"
+
+[dependencies]
+bee-common = { version = "0.5.0", path = "../../bee-common/bee-common" }
+bee-ledger = { version = "0.5.0", path = "../../bee-ledger" }
+bee-message = { version = "0.1.5", path = "../../bee-message" }
+bee-storage = { version = "0.9.0", path = "../bee-storage" }
+bee-tangle = { version = "0.2.0", path = "../../bee-tangle" }
+
+num_cpus = "1.13"
```
Comment (thibault-martinez) [inline bee-storage/bee-storage-volatile/Cargo.toml]: Add patches please

File: bee-storage/bee-storage-volatile/Cargo.toml [hunk 8]
```diff
@@ -0,0 +1,26 @@
+[package]
+name = "bee-storage-volatile"
```
Comment (thibault-martinez) [inline bee-storage/bee-storage-volatile/Cargo.toml]: Would it be better called `bee-storage-memory` ? Would better fit the definition you give of it IMO.

**Review by thibault-martinez — CHANGES_REQUESTED** [review]

File: bee-storage/bee-storage-memory/src/access/batch.rs [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+}
+
+impl BatchBuilder for Storage {
+    type Batch = StorageBatch;
+
+    fn batch_begin() -> Self::Batch {
+        Self::Batch::default()
+    }
+
+    fn batch_commit(&self, batch: Self::Batch, _durability: bool) -> Result<(), <Self as StorageBackend>::Error> {
```
Comment (thibault-martinez) [inline bee-storage/bee-storage-memory/src/access/batch.rs:56]: This doesn't look atomic ?

Comment (pvdrz) [reply bee-storage/bee-storage-memory/src/access/batch.rs:56]: It is atomic for each table, should it be atomic over the whole storage? I think the `sled` backend is not either.

Comment (thibault-martinez) [reply bee-storage/bee-storage-memory/src/access/batch.rs:56]: Yeah I just noticed it's not atomic in sled but it definitely should... That's the whole point of the batch

**Review by thibault-martinez — APPROVED** [review]

**Pull request merged**